Add Deep link validation tool  home page with fake data

hangyujin opened the pull request:
Work towards https://github.com/flutter/devtools/issues/6013.

 All work for the deep link validation tool will be hidden behind FeatureFlags.deepLinkValidation until the tool is ready to ship.


<img width="1264" alt="D785360C-BBB8-4AB1-AC6A-8E0D8F967CD2" src="https://github.com/flutter/devtools/assets/108393416/87c12cc0-7f88-4846-94b0-4eb87ecb7419">




![build.yaml badge]

If you need help, consider asking for help on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[build.yaml badge]: https://github.com/flutter/devtools/actions/workflows/build.yaml/badge.svg

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+    this.pathError = false,
+  });
+
+  String os;
+  String path;
+  String domain;
+  String scheme;
+  bool domainError;
+  bool pathError;
```
can these be final?

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
-
 import '../../shared/screen.dart';
 
+const List<String> paths = [
+  '/shoes/..*',
+  '/Clothes/..*',
+  '/Toys/..*',
+  '/Jewelry/..*',
+  '/Watches/..* ',
+  '/Glasses/..*',
+];
+
+List<LinkData> allLinkDatas = [
```
make these lists private and add dart doc to describe what they are for

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
-
 import '../../shared/screen.dart';
 
+const List<String> paths = [
```
nit: either remove type or put on the RHS:
`const _paths = <String>[ ... ];`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+    ),
+];
+
+class LinkData {
```
please add dartdoc

kenzieschmoll replied:
consider putting classes related to deep link data models in a different file `deep_links_model.dart` so that `deep_links_screen.dart` can consist of widget classes and UI code only.

Across devtools we generally keep UI classes (flutter widgets), model classes, and controller classes all in their separate categories. Not one class per file by any means, but grouping classes by UI, model, and controller helps with organization

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+  bool pathError;
+  String get searchLabel => (os + path + domain + scheme).toLowerCase();
+
+  DataRow buildRow(
```
instead of having a method that returns a widget, create a stateless widget `_LinkDataRow`. In general, always having a method that returns a widget is code smell that we should just create a stateless widget instead. This is cleaner stylistically and it makes it easier to debug widget trees with the inspector.

chunhtai replied:
agree, this seems like a data class, it would make more sense to move this method out of this class.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+        DataCell(Text(scheme)),
+        DataCell(
+          domainError
+              ? Row(
```
I would play with the screen size a bit and make sure this does not overflow. If it does, you'll likely need to use Flexible or Expanded around some of the children

hangyujin replied:
tried and it looks like DataTable widget already handles narrow screen well.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+                      Icons.error,
+                      color: Theme.of(context).colorScheme.error,
+                    ),
+                    const SizedBox(width: 10),
```
we have lots of consts you can reuse in `theme.dart` to ensure consistent styling across our tools. Use `denseSpacing` (8.0) here and elsewhere instead. We also have `defaultSpacing` with is 16.0. Those are likely the two most popular spacing consts you'll be using.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+        DataCell(Text(os)),
+        DataCell(Text(scheme)),
+        DataCell(
+          domainError
+              ? Row(
```
instead of using a ternary, you can conditionally add children to the list in Row:

```
Row(
  children: [
    if (domainError)
      Padding(
        padding: const EdgeInsets.only(right: denseSpacing),
        child: Icon (...),
      ),
    Text(domain),
  ],
)
```

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+        ),
+        DataCell(
+          pathError
+              ? Row(
```
same comment here about conditionally adding items to the Row's children based on the value of `pathError` rather than using a ternary

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
     );
   }
+
+  Widget buildDataTable(BuildContext context) {
```
make this a stateless widget class instead of having a method that returns a widget

hangyujin replied:
Updated it to use 

ValueListenableBuilder<List<LinkData>>(
      valueListenable: controller.linkDatasNotifier,

so  it is not a  a stateless widget  now.

kenzieschmoll replied:
While `ValueListenableBuilder` is not a stateless widget, a helper widget would still be stateless. In general, we never want to return a widget from a method. Always prefer a helper widget instead. 

```
class _DataTable extends StatelessWidget {
  ...
  
  Widget build(BuildContext context) {
    return ...;
  }
}
```

hangyujin replied:
done

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+  bool bundleByDomain = false;
+
+  void setBundleByDomain(bool shouldBundleByDomain) {
+    setState(() {
+      bundleByDomain = shouldBundleByDomain;
+    });
```
instead of using setState directly, I recommend creating a controller class with ValueNotifiers that the UI listens to. If you keep state in the DeepLinkPage state object directly, the state will be lost if the user switches to a separate screen or does something to cause a complete tear down and rebuild of the state object.

We do this for almost every screen in DevTools. Look where `DeepLinkPage` is created to see how other screens pass in a controller that will live for the entire lifecycle of DevTools. Then you can keep anything stateful in the controller object and never have to call set state directly or store data in the widget state.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+
+  @override
+  Widget build(BuildContext context) {
+    return ListView(
```
use a column instead and wrap the data table in an Expanded widget

hangyujin replied:
The data table can be longer than the screen height.

kenzieschmoll replied:
Do we really want the controls to scroll out of view if the data table is large though? I suggest using a column and then making the data table itself scrollable so that we do not lose the controls to scroll through the table. You can see an example of this on other DevTools pages like the logging page, for example, where the controls at the top of the screen are fixed and only the table scrolls.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+            ),
+          ],
+        ),
+        const SizedBox(height: 10),
```
use denseSpacing or defaultSpacing

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+              },
+              child: const Text('Bundle by domain'),
+            ),
+            const SizedBox(width: 10),
```
use denseSpacing or defaultSpacing

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+
+Widget buildTitle(String text, BuildContext context) {
+  final textTheme = Theme.of(context).textTheme;
+  return Text(
```
declare this Text widget directly in the build method it is used in above instead of having a method that returns a widget. This is too simple to warrant creating its own stateless widget class

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+        DataColumn(label: Text('Domain')),
+        DataColumn(label: Text('Path')),
+      ],
+      dataRowMinHeight: bundleByDomain ? 150 : null,
```
please put 150 in a named const and use that here and below

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+      for (var i = 0; i < linkDatas.length; i++)
+        linkDatas[i].buildRow(
+          context,
+          color: i % 2 == 0
+              ? MaterialStateProperty.all<Color>(
+                  Theme.of(context).colorScheme.onSurface.withOpacity(0.2),
+                )
+              : null,
```
use `alternatingColorForIndex` from `common_widgets.dart`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+              child: const Text('Bundle by domain'),
+            ),
+            const SizedBox(width: 10),
+            SearchBar(
```
may be too advanced for the current use case, but consider using `SearchField` from `search.dart` in DevTools, which has advanced functionality to hook search into your screen controller and provides some advanced features like navigation through results.

hangyujin replied:
SearchableFlatTable seems to highlight the result but I want to filter the data with the search content.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +130,112 @@ class DeepLinksScreen extends Screen {
+                  searchContent = value.toLowerCase();
+                });
+              },
+              constraints: BoxConstraints.tight(const Size(200, 40)),
```
please use `defaultSearchFieldWidth` and `defaultTextFieldHeight` from `theme.dart`. Many consts in theme.dart are also wrapped with `scaleByFontFactor` which ensures your UI will scale nicely as the font factor changes

kenzieschmoll reviewed (commented):
done with first pass

chunhtai commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+    return LinkData(
+      os: os,
+      domain: domain,
+      path: '${linkData.path}\n$path',
```
seems weird to use an arbitrary delimiter, why not keep a list instead?

chunhtai commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -3,9 +3,115 @@
+
+    return LinkData(
+      os: os,
+      domain: domain,
```
does the domain need to be the same?

hangyujin replied:
yes

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +28,138 @@ class DeepLinksScreen extends Screen {
   }
 }
+
+DataRow buildRow(
```
same here. Make this a helper widget instead of having a method that returns a widget

hangyujin replied:
the `DataRow` is not a widget, it's a data structure used in `DataTable`

chunhtai commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_model.dart:
```diff
@@ -0,0 +1,37 @@
+    if (linkData == null) {
+      return this;
+    }
+
```
can you add an assert that the domain is the same?

hangyujin replied:
done

chunhtai commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -4,7 +4,11 @@
+import 'deep_links_controller.dart';
+import 'deep_links_model.dart';
+
+const bundledDataRowHeight = 150.0;
```
```suggestion
const kBundledDataRowHeight = 150.0;
```

chunhtai commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:
```diff
@@ -2,4 +2,51 @@
+        : allLinkDatas;
+
+    if (bundleByDomain) {
+      final Map<String, LinkData> bundleByDomainMap = {};
```
```suggestion
      final bundleByDomainMap = <String, LinkData>{};
```
same for other map initialization

chunhtai replied:
@kenzieschmoll is there a lint for this?

kenzieschmoll replied:
Not that I'm aware of. CC @pq.

pq replied:
[`omit_local_variable_types`](https://dart.dev/tools/linter-rules/omit_local_variable_types) handles this case but it may be a bridge too far... Check it out?

chunhtai reviewed (approved):
LGTM if also LGT @kenzieschmoll , left some nit picks

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:
```diff
@@ -2,4 +2,51 @@
+
+class DeepLinksController {
+
+  final _linkDatasNotifier = ValueNotifier<List<LinkData>>(allLinkDatas);
+  ValueListenable<List<LinkData>> get linkDatasNotifier => _linkDatasNotifier;
```
style nit: order getters before private variables: https://github.com/flutter/devtools/blob/master/packages/README.md#order-of-getters-and-setters. Here and elsewhere

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -4,7 +4,11 @@
 import 'package:flutter/material.dart';
 
-import '../../shared/screen.dart';
+import '../../../devtools_app.dart';
```
do not import devtools_app.dart directly inside of `lib/`. This is fine to import in tests, but inside lib we should import the libraries we need directly instead of importing this large file that exports many files.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -24,8 +28,144 @@ class DeepLinksScreen extends Screen {
+  @override
+  Widget build(BuildContext context) {
+    final List<DataRow> rows = [
+      for (var i = 0; i < linkDatas.length; i++)
```
will `linkDatas` be large? We should be careful about building the entire table if there will be a lot of data here, as this could cause performance issues. If the data set is large, we should build this table in a lazy way.

hangyujin replied:
Thank you! I've updated it to use the FlatTable

chunhtai replied:
some large company may have ~1000 deeplink.

hangyujin commented:
Updated UI to match the new mock 
<img width="1257" alt="20D8B5A1-B956-4DC5-89FC-2D582CD5C14F" src="https://github.com/flutter/devtools/assets/108393416/996ce3ff-a3e8-426a-8c61-a20e2a3edee7">

<img width="1275" alt="2C8D31DA-4278-42C1-A7A1-FBCA73E5D8D8" src="https://github.com/flutter/devtools/assets/108393416/1d990127-9af9-46a3-90e3-984c46f0dd91">

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:
```diff
@@ -7,46 +7,63 @@ import 'package:flutter/foundation.dart';
   }
+
+  List<LinkData> get getLinkDatasByPath {
+    final Map<String, LinkData> linkDatasByPath = {};
```
nit: here and for the method below, put types on the RHS: `final linkDatasByPath = <String, LinkData>{};`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_model.dart:
```diff
@@ -3,37 +3,180 @@
+          Padding(
+            padding: const EdgeInsets.only(right: denseSpacing),
+            child: Icon(
+              Icons.error,
```
add `size: defaultIconSize`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_model.dart:
```diff
@@ -3,37 +3,180 @@
+              Icons.error,
+              color: Theme.of(context).colorScheme.error,
+            ),
+          ),
```
consider adding an `else` statement that adds a `SizedBox()` of the same width as the total Padding + Icon width, so that the table rows do not have staggered indents depending on whether there is an error or not

hangyujin replied:
In our design mock, icons are aligned together with text like 
<img width="471" alt="5D8DC136-CBC2-4EAC-A327-7C366B0A7817" src="https://github.com/flutter/devtools/assets/108393416/6e867f5c-cdf6-44fd-8732-01fc7b2ff104">

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_model.dart:
```diff
@@ -3,37 +3,180 @@
-      domainError: domainError || linkData.domainError,
+  @override
+  Widget getDisplayWidget(LinkData dataObject, BuildContext context) {
+    return Row(
+      children: [
+        if (dataObject.pathError)
+          Padding(
+            padding: const EdgeInsets.only(right: denseSpacing),
+            child: Icon(
+              Icons.error,
+              color: Theme.of(context).colorScheme.error,
+            ),
+          ),
+        const SizedBox(width: denseSpacing),
+        Text(
+          dataObject.path,
+          overflow: TextOverflow.ellipsis,
+        ),
+      ],
```
This is repeated above. Please create a helper widget `_ErrorAwareText` or similar that you can use in both DomainColumn and PathColumn

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_model.dart:
```diff
@@ -3,37 +3,180 @@
+    } else {
+      return const Text(
+        'No issues found',
+        style: TextStyle(color: Color.fromARGB(255, 156, 233, 195)),
```
can we use a color from `theme.dart` instead of creating a new one here?

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
-                'All deep links',
-                style: Theme.of(context).textTheme.bodyLarge,
+    return DefaultTabController(
+      length: 3,
```
Use `TableView.values.length`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:
```diff
@@ -7,46 +7,63 @@ import 'package:flutter/foundation.dart';
+
+  set showSpitScreen(bool value) {
+    _showSpitScreenNotifier.value = value;
   }
```
If we are going to have a public getter and setter for a notifier, then there is no need to have a private notifier and public ValueListenable. So for this one, I reccomend we just make the `_showSplitScreenNotifier` public

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_controller.dart:
```diff
@@ -7,46 +7,63 @@ import 'package:flutter/foundation.dart';
-  final _linkDatasNotifier = ValueNotifier<List<LinkData>>(allLinkDatas);
   ValueListenable<List<LinkData>> get linkDatasNotifier => _linkDatasNotifier;
+  ValueListenable<String> get searchContentNotifier => _searchContentNotifier;
+  ValueListenable<bool> get showSpitScreenNotifier => _showSpitScreenNotifier;
```
style nit: order getters, private members, and setters per the guidelines here: https://github.com/flutter/devtools/blob/master/packages/README.md#order-of-getters-and-setters

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          Row(
```
Can you try to use `AreaPaneHeader` for consistent styling with the rest of DevTools?

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
-              constraints: BoxConstraints.tight(
-                Size(defaultSearchFieldWidth, defaultTextFieldHeight),
+              const SizedBox(width: denseSpacing),
+              SearchBar(
```
Are you able to use `SearchField` from `search.dart`? There are a few examples of this throughout DevTools, but this would give you a search field styled consistently with the rest of DevTools that supports navigation through search results.

hangyujin replied:
is  SearchField paired with  SearchableFlatTable?

kenzieschmoll replied:
They are not tightly coupled no. They can be used together (like for the CPU profiler Method Table, for example), but you can also use SearchField independently. It just requires a search controller, so you'l make your deep links controller mixin SearchControllerMixin for that. If you look at usages of `SearchField` in DevTools, you'll see several examples of using the `SearchField` widget with a controller that mixes in `SearchControllerMixin`. The logging screen and logging controller are an example of this.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+            ],
+          ),
+          const SizedBox(height: denseSpacing),
+          const TabBar(
```
consider using `AnalyticsTabbedView` so that you will automatically get analytics reported for tab switches. This would be good information to know which tab users are using the most.

kenzieschmoll replied:
this comment still stands as well

hangyujin replied:
I think I will add analytics in a separated PR

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+    return Row(
+      children: [
+        Expanded(
+          child: _DataTable(
```
wrap with `RoundedOutlinedBorder`. To make this look right with the header, you may need to play with the `show*` parameters.

hangyujin replied:
Will update border later according to figma

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+          Expanded(
+            child: ValueListenableBuilder<LinkData?>(
+              valueListenable: controller.selectedLink,
+              builder: (context, selectedLink, _) => _ValidationDetailScreen(
```
wrap this widget with a `RoundedOutlinedBorder` also

hangyujin replied:
Will update border later according to figma

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -4,11 +4,21 @@
 import 'deep_links_model.dart';
 
-const bundledDataRowHeight = 150.0;
+enum TableView {
```
nit: maybe rename to `TableViewType` to make it more clear that this is a type enum and not a widget? When reading the code below,`TableView` sounded like a widget type

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
-        DataColumn(label: Text('Scheme')),
-        DataColumn(label: Text('Domain')),
-        DataColumn(label: Text('Path')),
+    return FlatTable(
```
If you use `SearchableFlatTable` instead of `FlatTable` you should be able to get the search functionality you added above for free. You'll just have to make your `controller` mix in `SearchControllerMixin`

hangyujin replied:
I tried that and SearchableFlatTable highlights the search result, and I need to filter the result.

hangyujin replied:
I can use DevToolsClearableTextField, but I am still using `FlatTable` not `SearchableFlatTable` as this example is also using `FlatTable`

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+            ' file on your website. App checks are done for the intent filters in'
+            ' the manifest and info.plist file, routing issues, URL format, etc.',
+            style: Theme.of(context).textTheme.bodySmall,
+          ),
```
Consider adding a `PaddedDivider` widget here before 'Domain check'

hangyujin replied:
There is no divider in figma so I will follow that.

kenzieschmoll replied:
Okay. Generally speaking though @nancyhu2023, does the design file in Figma use the same color scheme / style practices as the remainder of DevTools? Since the Deep Link tool is part of DevTools, I think it is important that we have some consistency with the design patterns used here and those used throughout the remainder of the Devtools tool suite.

kenzieschmoll commented on packages/devtools_app/lib/src/screens/deep_link_validation/deep_links_screen.dart:
```diff
@@ -52,119 +62,251 @@ class _DeepLinkPageState extends State<DeepLinkPage>
+            style: Theme.of(context).textTheme.bodySmall,
+          ),
+          const Text('Domain check'),
+          _DomainCheckTable(linkData: linkData),
```
wrap in Expanded

kenzieschmoll commented on packages/devtools_app/lib/src/shared/table/table_data.dart:
```diff
@@ -79,6 +79,8 @@ abstract class ColumnData<T> {
   /// Get the cell's display value from the given [dataObject].
   String getDisplayValue(T dataObject) =>
       getValue(dataObject)?.toString() ?? '';
+  
+  Widget? getDisplayWidget(T dataObject, BuildContext context) => null ;
```
Delete this and have your columns implement `ColumnRenderer` instead. Example: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/logging/_kind_column.dart#L13

kenzieschmoll commented on packages/devtools_app/lib/src/shared/table/table.dart:
```diff
@@ -1486,28 +1486,29 @@ class _TableRowState<T> extends State<TableRow<T>>
-                  style: const TextStyle(
-                    fontStyle: FontStyle.italic,
-                  ),
+        content ??= column.getDisplayWidget(node, context) ??
```
once you have your columns implement `ColumnRenderer`, please revert changes to this file

kenzieschmoll reviewed (commented):
another round of review. Looking good!

hangyujin commented:
Hi @kenzieschmoll, I addressed most of comments but I am not sure about whether to use SearchableFlatTable and SearchField.

Also can I merge this PR as a version 0 and  consider updating it to use SearchableFlatTable later to unblock Chun-Heng to test his backend changes on top of this simple UI?

kenzieschmoll reviewed (approved):
Looks like you'll need to merge with master and resolve a merge conflict, but lgtm!
